### PR TITLE
Adding 2-hidden layer neural network with back propagation using sigmoid activation function 

### PR DESCRIPTION
**Pull request by raman77768**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by cclauss** [comment]
@dhruvmanila Your review, please?

**Comment by dhruvmanila** [comment]
@algorithms-keeper review

File: neural_network/2_hidden_layers_neural_network.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
+import numpy
+
+
+class NeuralNetwork:
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:4]: Please provide a better class name here. What type of neural network is this class representing?

File: neural_network/2_hidden_layers_neural_network.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
+import numpy
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:1]: Conventions say to import and use numpy as: `import numpy as np`

File: neural_network/2_hidden_layers_neural_network.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+
+class NeuralNetwork:
+    def __init__(self, input_array: numpy.array, output_array: numpy.array) -> None:
+        """
+        input_array : input values for training the neural network.
+        output_array : expected output values of the given inputs.
+        """
+        self.input = input_array
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:10]:
```suggestion
        self.input_array = input_array
```

File: neural_network/2_hidden_layers_neural_network.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        # Initial weights are assigned randomly where first argument is the number
+        # of nodes in previous layer and second argument is the
+        # number of nodes in the next layer.
+
+        # random initial weights for the input layer
+        # self.input.shape[1] is used to represent number of nodes in input layer
+        # first hidden layer consists of 4 nodes
+        self.weights1 = numpy.random.rand(self.input.shape[1], 4)
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:18]: Instead of weights1, weights2,... we should provide descriptive names such that when we're referencing them in other parts of the class they are easy to pick up. Maybe it should be named: `input_layer_weights, first_layer_weights, etc`. You can change it to whatever suits you best.

File: neural_network/2_hidden_layers_neural_network.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        # first hidden layer has 4 nodes
+        # second hidden layer has 3 nodes
+        self.weights2 = numpy.random.rand(4, 3)
+
+        # random initial weights for the second hidden layer
+        # second hidden layer has 3 nodes
+        # output layer has 1 node
+        self.weights3 = numpy.random.rand(3, 1)
+
+        self.y = output_array
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:30]: What is `y`? Please make it more descriptive.

File: neural_network/2_hidden_layers_neural_network.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        >>> nn = NeuralNetwork(input_val,output_val)
+        >>> res = nn.feedforward()
+        >>> array_sum = numpy.sum(res)
+        >>> array_has_nan = numpy.isnan(array_sum)
+        >>> print(array_has_nan)
+        False
+        """
+        # layer1 is the layer connecting the input nodes with
+        # the first hidden layer nodes
+        self.layer1 = sigmoid(numpy.dot(self.input, self.weights1))
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:49]: Same as the comment for `weights1, weights2, ...`
Descriptive names for `layer1, layer2, ...`

File: neural_network/2_hidden_layers_neural_network.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            self.output = self.feedforward()
+            self.back_propagation()
+            if give_loss:
+                print(
+                    "Iteration %s " % iteration,
+                    "Loss: "
+                    + str(numpy.mean(numpy.square(output - self.feedforward()))),
+                )
+
+    def predict(self, input: list) -> int:
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:123]: Please provide the list elements type: `List[type]` using `from typing import List`

File: neural_network/2_hidden_layers_neural_network.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        0
+        """
+        self.array = input
+        self.layer1 = sigmoid(numpy.dot(self.array, self.weights1))
+        self.layer2 = sigmoid(numpy.dot(self.layer1, self.weights2))
+        self.layer3 = sigmoid(numpy.dot(self.layer2, self.weights3))
+        if self.layer3 > 0.45:
+            return 1
+        else:
+            return 0
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:141]: What does the output values represent? Please mention this in the function contract (function docstring).

1 and 0 represent Truthy and Falsy values, so if you're going for that then you should probably return a boolean value.

File: neural_network/2_hidden_layers_neural_network.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    """
+    returns derivative of the sigmoid value
+
+    >>> sigmoid_derivative(0.7)
+    0.22171287329310904
+    """
+    return sigmoid(value) * (1 - sigmoid(value))
+
+
+def example() -> int:
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:168]: Please explain what this example is about in the function docstring.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
Looks like a great start but there are things that need to be taken care of:
- Provide a brief description regarding the function contract: what the function is about, what it does, what the user can expect while using the function.
- Use proper sentences while writing which means each sentence should begin with an uppercase letter and end with a period.
- While writing, imagine that your program is being studied by some random person, will he be able to deduce what this program is about directly from the code, comments, and docstring?

File: neural_network/2_hidden_layers_neural_network.py (hunk 2 above)
Comment (cclauss) [reply neural_network/2_hidden_layers_neural_network.py:1]: I do not think that we need to enforce this.

**Comment by raman77768** [comment]
@dhruvmanila  Made the changes, Please have a look and let me know.

File: neural_network/2_hidden_layers_neural_network.py (hunk 2 above)
Comment (dhruvmanila) [reply neural_network/2_hidden_layers_neural_network.py:3]: It's not about enforcing, it's about being consistent as I think all the scripts in this repository is using `numpy` like so.

File: neural_network/2_hidden_layers_neural_network.py (hunk 8 above)
Comment (dhruvmanila) [reply neural_network/2_hidden_layers_neural_network.py:212]: What I mean by the above comment is to mention what does 0 and 1 represent. If the user uses the function and gets 0 and 1, how does he/she know what to do with that?

Also, can you tell me why should we not return the value itself which is between 0 and 1 instead?

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
This is looking good! Thanks for all the efforts! Can you provide some links to the references used and also links which points the user to understand more about the topics used here?

File: neural_network/2_hidden_layers_neural_network.py (hunk 8 above)
Comment (raman77768) [reply neural_network/2_hidden_layers_neural_network.py:212]: The output is given in binary classification, '0' and '1' are the two classes provided in this training data.Our model provides a value in-between 0-1 so i have set a threshold value which assigns the value to one of the classes

**Comment by dhruvmanila** [comment]
@raman77768 Please resolve the conversation for which the changes have been implemented, that will make the review process easier.

**Comment by raman77768** [comment]
@dhruvmanila Resolved the conversations where changes were made.

File: neural_network/2_hidden_layers_neural_network.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        >>> nn = NeuralNetwork(input_val,output_val)
+        >>> nn.train(output_val,100,False)
+        >>> nn.predict([0,0,0])
+        0
+        """
+        self.array = input
+        self.layer1 = sigmoid(numpy.dot(self.array, self.weights1))
+        self.layer2 = sigmoid(numpy.dot(self.layer1, self.weights2))
+        self.layer3 = sigmoid(numpy.dot(self.layer2, self.weights3))
+        if self.layer3 > 0.45:
```
Comment (cclauss) [inline neural_network/2_hidden_layers_neural_network.py]:
```suggestion
        int(self.layer3 > 0.45)
```

**Review by cclauss — APPROVED** [review]
Looks good to me.  Thanks for your persistence!!

File: neural_network/2_hidden_layers_neural_network.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        self.second_hidden_layer_and_output_layer_weights = numpy.random.rand(3, 1)
+
+        # Real output values provided.
+        self.output_array = output_array
+
+        # Predicted output values by the neural network.
+        # Predicted_output array initially consists of zeroes.
+        self.predicted_output = numpy.zeros(output_array.shape)
+
+    def feedforward(self) -> None:
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:54]: This function has a return statement but the return type is `None`.

File: neural_network/2_hidden_layers_neural_network.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        error rate obtained in the previous epoch (i.e., iteration).
+        Updation is done using derivative of sogmoid activation function.
+
+        >>> input_val = numpy.array(([0, 0, 0], [0, 0, 0], [0, 0, 0]), dtype=float)
+        >>> output_val = numpy.array(([0], [0], [0]), dtype=float)
+        >>> nn = TwoHiddenLayerNeuralNetwork(input_val, output_val)
+        >>> res = nn.feedforward()
+        >>> nn.back_propagation()
+        >>> updated_weights = nn.second_hidden_layer_and_output_layer_weights
+        >>> (res == updated_weights).all()
```
Comment (dhruvmanila) [inline neural_network/2_hidden_layers_neural_network.py:109]: Can you tell me what this expression means? I know the builtin `all` but it's probably not supposed to be used like this, so I am a bit confused.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
`mypy` still complains:
```console
$ mypy 2_hidden_layers_neural_network.py
2_hidden_layers_neural_network.py:95: error: No return value expected
2_hidden_layers_neural_network.py:114: error: "float" has no attribute "T"
2_hidden_layers_neural_network.py:120: error: "float" has no attribute "T"
2_hidden_layers_neural_network.py:179: error: "feedforward" of "TwoHiddenLayerNeuralNetwork" does not return a value
2_hidden_layers_neural_network.py:182: error: "feedforward" of "TwoHiddenLayerNeuralNetwork" does not return a value
```

File: neural_network/2_hidden_layers_neural_network.py (hunk 12 above)
Comment (raman77768) [reply neural_network/2_hidden_layers_neural_network.py:109]: The "res" variable contains the weights after first feedforward in neural network and "updated_weights" contains the array after backpropagation, so the (res == updated_weights).all() lines check whether the weights were changed or not i.e. compares the elements of both the arrays and return True/False.
So we are check if our back_propagtion function is updating the weights or not , if weights are not updated then (res == updated_weights).all() will return True i.e. all the weights are still same after back propagation.

**Comment by dhruvmanila** [comment]
@raman77768 Please run `mypy` locally and fix all the errors. If you have any doubt, feel free to ask.

**Comment by raman77768** [comment]
image

@dhruvmanila 
error: "float" has no attribute "T", Here i am using .T(i.e. transpose on numpy array) from numpy library but it is stating them as float values

**Comment by dhruvmanila** [comment]
That is because you provided the incorrect type hint for the parameter and for the function `sigmoid` and also `sigmoid_derivative`. You are stating that it returns a `float` value, thus the variable became float and when you try to access `T`, mypy complains correctly.

Another thing to note is that `numpy` does not provide type hints, so work accordingly.

**Review by dhruvmanila — APPROVED** [review]
Awesome work! LGTM! 🎉 
Thanks for the patience and persistence! Appreciate it! 👍

**Comment by cclauss** [comment]
Yeah!!  64 comments!  Is that a record?

**Comment by dhruvmanila** [comment]
Lol, a record for what? Also, some of the reviews were made by the bot ;-)
